DependencyInjection/SwiftmailerTransportFactory.php

```diff
@@ -0,0 +1,138 @@
+
+use Symfony\Component\Routing\RequestContext;
+
+/**
+ * Factory to create a \Swift_Transport object.
+ *
+ * @author Romain Gautier <mail@romain.sh>
+ */
+class SwiftmailerTransportFactory
+{
+    /**
+     * @param array                         $options
+     * @param RequestContext                $requestContext
+     * @param \Swift_Events_EventDispatcher $eventDispatcher
+     *
+     * @return \Swift_Transport
+     *
+     * @throws \InvalidArgumentException if the scheme is not a built-in Swiftmailer transport
+     */
+    public static function createTransport(array $options, RequestContext $requestContext, \Swift_Events_EventDispatcher $eventDispatcher)
+    {
+        $transport = static::resolveOptions($options);
+
+        if ('smtp' === $transport) {
+            $smtpAuthHandler = new \Swift_Transport_Esmtp_AuthHandler([
+                new \Swift_Transport_Esmtp_Auth_CramMd5Authenticator(),
+                new \Swift_Transport_Esmtp_Auth_LoginAuthenticator(),
+                new \Swift_Transport_Esmtp_Auth_PlainAuthenticator(),
+            ]);
+            $smtpAuthHandler->setUsername($options['user']);
+            $smtpAuthHandler->setPassword($options['pass']);
+            $smtpAuthHandler->setAuthMode($options['auth_mode']);
+
+            $transport = new \Swift_Transport_EsmtpTransport(
+                new \Swift_Transport_StreamBuffer(new \Swift_StreamFilters_StringReplacementFilterFactory()),
+                array($smtpAuthHandler),
+                $eventDispatcher
+            );
+            $transport->setHost($options['host']);
+            $transport->setPort($options['port']);
+            $transport->setEncryption($options['encryption']);
+            $transport->setTimeout($options['timeout']);
+            $transport->setSourceIp($options['source_ip']);
+
+            (new SmtpTransportConfigurator(null, $requestContext))->configure($transport);
+        } elseif ('sendmail' === $transport) {
+            $transport = new \Swift_Transport_SendmailTransport(
+                new \Swift_Transport_StreamBuffer(new \Swift_StreamFilters_StringReplacementFilterFactory()),
+                $eventDispatcher
+            );
+
+            (new SmtpTransportConfigurator(null, $requestContext))->configure($transport);
+        } elseif ('mail' === $transport) {
+            $transport = new \Swift_Transport_MailTransport(
+                new \Swift_Transport_SimpleMailInvoker(),
+                $eventDispatcher
+            );
+        } elseif ('null' === $transport) {
+            $transport = new \Swift_Transport_NullTransport($eventDispatcher);
+        } else {
+            throw new \InvalidArgumentException(sprintf('Not a built-in Swiftmailer transport: %s.', $options['dsn']));
+        }
+
+        return $transport;
+    }
+
+    /**
+     * @param array $options
+     *
+     * @return string transport
+     */
+    public static function resolveOptions(array &$options)
```

Comment:
instead of passing this by reference, I'd make the function return $options, with `$options['transport']` always set